### ISSUE TITLE
Unable to acquire dns-proxy lock and high CPU-usage

### ISSUE THREAD

**Issue by mhulscher** [cilium/cilium#27156]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Several of my cilium-agents are using a lot of CPU: ~2 CPU. All of these agents are complaining about not being able to acquire a L7-proxy lock and not being able to apply L7-policy changes.

### Cilium Version

1.14.0

### Kernel Version

5.10.184-175.731.amzn2.aarch64

### Kubernetes Version

v1.26.6-eks-a5565ad

### Sysdump

[cilium-sysdump-20230731-134938.zip](https://github.com/cilium/cilium/files/12214674/cilium-sysdump-20230731-134938.zip)
[cilium-sysdump-20230731-135423.zip](https://github.com/cilium/cilium/files/12214685/cilium-sysdump-20230731-135423.zip)
[cilium-sysdump-20230731-140512.zip](https://github.com/cilium/cilium/files/12214806/cilium-sysdump-20230731-140512.zip)

### Relevant log output

```shell
cilium-8fcq7 cilium-agent level=warning msg="Lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --dnsproxy-lock-count and/or --dnsproxy-lock-timeout" dnsName=loki-distributor.loki.svc.cluster.local. duration=628.69381ms expected=500ms subsys=daemon
cilium-8fcq7 cilium-agent level=warning msg="Failed to apply L7 proxy policy changes. These will be re-applied in future updates." error="context cancelled before waiting for proxy updates: context deadline exceeded" subsys=endpoint-manager
cilium-8fcq7 cilium-agent level=error msg="Timed out waiting for datapath updates of FQDN IP information; returning response" subsys=daemon
cilium-8fcq7 cilium-agent level=error msg="Timed out waiting for datapath updates of FQDN IP information; returning response" subsys=daemon
cilium-8fcq7 cilium-agent level=warning msg="Lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --dnsproxy-lock-count and/or --dnsproxy-lock-timeout" dnsName=loki-query-scheduler.loki.svc.cluster.local. duration=692.425034ms expected=500ms subsys=daemon
cilium-8fcq7 cilium-agent level=error msg="Timed out waiting for datapath updates of FQDN IP information; returning response" subsys=daemon
cilium-8fcq7 cilium-agent level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
cilium-8fcq7 cilium-agent level=warning msg="Failed to apply L7 proxy policy changes. These will be re-applied in future updates." error="context cancelled before waiting for proxy updates: context deadline exceeded" subsys=endpoint-manager
cilium-8fcq7 cilium-agent level=warning msg="Lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --dnsproxy-lock-count and/or --dnsproxy-lock-timeout" dnsName=aio-f7ba-tempo.s3.dualstack.eu-west-1.amazonaws.com. duration=929.900205ms expected=500ms subsys=daemon
cilium-8fcq7 cilium-agent level=warning msg="Lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --dnsproxy-lock-count and/or --dnsproxy-lock-timeout" dnsName=aio-f7ba-tempo.s3.dualstack.eu-west-1.amazonaws.com. duration=943.71616ms expected=500ms subsys=daemon
cilium-8fcq7 cilium-agent level=info msg="regenerating all endpoints" reason= subsys=endpoint-manager
```

### Anything else?

The affected pods are:

* cilium-8fcq7
* cilium-ll7tb
* cilium-29bhp

![image](https://github.com/cilium/cilium/assets/18721288/b84c708e-332f-4079-96d3-66522bb89cc8)

Both coredns and hubble l7/dns-metrics seem to suggest that we are doing approx 15 dns queries per second, which seems very low. There is no CLRP because of #27084 .

![image](https://github.com/cilium/cilium/assets/18721288/8ffec59e-5b1f-43fb-b3b7-347fa26a4f97)

These are my helm values:

```
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      - matchExpressions:
        - key: eks.amazonaws.com/compute-type
          operator: NotIn
          values:
          - fargate
agentNotReadyTaintKey: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-not-ready
bpf:
  masquerade: false
  tproxy: true
cluster:
  name: aio-f7ba
cni:
  chainingMode: none
dnsProxy:
  endpointMaxIpPerHostname: 1000
  minTtl: 0
enableIPv4Masquerade: false
encryption:
  enabled: true
  nodeEncryption: true
  type: wireguard
endpointHealthChecking:
  enabled: false
endpointStatus:
  enabled: true
  status: policy
eni:
  awsEnablePrefixDelegation: true
  ec2APIEndpoint: ec2.eu-west-1.amazonaws.com
  enabled: true
  eniTags: {}
  iamRole: arn:aws:iam::630765193817:role/aio-f7ba-cilium-operator
  instanceTagsFilter: aws:eks:cluster-name=aio-f7ba
  updateEC2AdapterLimitViaAPI: true
envoy:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: eks.amazonaws.com/compute-type
            operator: NotIn
            values:
            - fargate
  enabled: true
  nodeSelector:
    kubernetes.io/os: linux
  priorityClassName: system-node-critical
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
      labels:
        system: "true"
  resources:
    requests:
      cpu: 100m
      memory: 200Mi
  rollOutPods: true
  tolerations:
  - operator: Exists
  updateStrategy:
    type: OnDelete
extraConfig:
  api-rate-limit: |
    {"endpoint-create":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-delete":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-get":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-list":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true","endpoint-patch":"rate-limit:200/s,rate-burst:200,parallel-requests:200,auto-adjust:true"}
healthChecking: false
hubble:
  enabled: true
  eventBufferCapacity: "8191"
  metrics:
    enableOpenMetrics: true
    enabled:
    - dns:query;labelsContext=source_namespace,source_workload
    - httpV2:exemplars=true;labelsContext=source_namespace,source_workload,source_app,destination_namespace,destination_workload,destination_app,traffic_direction
    serviceMonitor:
      enabled: true
      labels:
        system: "true"
  relay:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: k8s-app
              operator: In
              values:
              - hubble-relay
          topologyKey: topology.kubernetes.io/zone
    enabled: true
    podDisruptionBudget:
      enabled: true
      maxUnavailable: 1
    replicas: 3
    resources:
      requests:
        cpu: 25m
    rollOutPods: true
    updateStrategy:
      rollingUpdate:
        maxSurge: 1
        maxUnavailable: 1
      type: RollingUpdate
  ui:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
            - key: k8s-app
              operator: In
              values:
              - hubble-ui
          topologyKey: topology.kubernetes.io/zone
    enabled: true
    ingress:
      className: ingress-nginx
      enabled: true
      hosts:
      - hubble.example.com
    podDisruptionBudget:
      enabled: true
      maxUnavailable: 1
    replicas: 2
    resources:
      requests:
        cpu: 25m
    rollOutPods: true
    updateStrategy:
      rollingUpdate:
        maxSurge: 1
        maxUnavailable: 1
      type: RollingUpdate
ipam:
  mode: eni
k8sServiceHost: 2863F87F2DC378B306123C09AAE448E5.gr7.eu-west-1.eks.amazonaws.com
k8sServicePort: 443
kubeProxyReplacement: true
l7Proxy: true
labels: k8s:!job-name k8s:!controller-uid
loadBalancer:
  l7:
    backend: envoy
  serviceTopology: true
localRedirectPolicy: true
operator:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: name
            operator: In
            values:
            - cilium-operator
        topologyKey: topology.kubernetes.io/zone
  extraArgs:
  - --unmanaged-pod-watcher-interval=0
  podDisruptionBudget:
    enabled: true
    maxUnavailable: 1
  priorityClassName: system-cluster-critical
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
      labels:
        system: "true"
  replicas: 2
  resources:
    requests:
      cpu: 25m
  rollOutPods: true
  tolerations:
  - key: node.kubernetes.io/not-ready
    operator: Exists
  - key: ignore-taint.cluster-autoscaler.kubernetes.io/cilium-not-ready
    operator: Exists
  - key: ignore-taint.cluster-autoscaler.kubernetes.io/dns-not-ready
    operator: Exists
pmtuDiscovery:
  enabled: true
policyEnforcementMode: default
priorityClassName: system-node-critical
prometheus:
  enabled: true
  serviceMonitor:
    enabled: true
    labels:
      system: "true"
resources:
  requests:
    cpu: 25m
svcSourceRangeCheck: "false"
tunnel: disabled
updateStrategy:
  type: OnDelete
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by christarazi** [comment]
Could you try with https://github.com/cilium/cilium/pull/27147?

**Comment by mhulscher** [comment]
@christarazi I tested with `level=info msg="  --dnsproxy-lock-count='131'" subsys=daemon` but unfortunately this doesn't solve the problem. No changes.

**Comment by mhulscher** [comment]
After some investigation, this seems to be caused by excessive entries S3-hostname entries in the fqdn cache. I notice that the fqdn cache frequently has more than 3k entries for a single S3-bucket hostname. I imagine this causes a lot of thrashing in the fqdn cahe. This can in part be explained:
* I configured the s3-entry to use the dualstack-variant, ie `aio-ehun-tempo.s3.dualstack.eu-west-1.amazonaws.com.`
* There are loads of ipv6 DNS responses in the fqdn cache
* I applied a wrong S3-configuration to my grafana-stack applications which causes them to create more connections than usual (retries to fetch/store objects)
* There is no node-local-dns at the moment so many more original DNS requests are made

I am going to get rid of the dualstack s3-endpoint and I am going to fix my components' s3 misconfiguration. I will report with the results afterwards.
